src/backend/distributed/transaction/lock_graph.c

```diff
@@ -0,0 +1,769 @@
+	waitEdge->waitingTransactionStamp = ParseTimestampTzField(result, rowIndex, 3);
+	waitEdge->blockingPid = ParseIntField(result, rowIndex, 4);
+	waitEdge->blockingNodeId = ParseIntField(result, rowIndex, 5);
+	waitEdge->blockingTransactionNum = ParseIntField(result, rowIndex, 6);
+	waitEdge->blockingTransactionStamp = ParseTimestampTzField(result, rowIndex, 7);
+	waitEdge->isBlockingXactWaiting = ParseBoolField(result, rowIndex, 8);
+}
+
+
+/*
+ * ParseIntField parses a int64 from a remote result or returns 0 if the
+ * result is NULL.
+ */
+static int64
+ParseIntField(PGresult *result, int rowIndex, int colIndex)
+{
+	char *resultString = NULL;
+
+	if (PQgetisnull(result, rowIndex, colIndex))
+	{
+		return 0;
+	}
+
+	resultString = PQgetvalue(result, rowIndex, colIndex);
+
+	return pg_strtouint64(resultString, NULL, 10);
+}
+
+
+/*
+ * ParseBoolField parses a bool from a remote result or returns false if the
+ * result is NULL.
+ */
+static bool
+ParseBoolField(PGresult *result, int rowIndex, int colIndex)
+{
+	char *resultString = NULL;
+
+	if (PQgetisnull(result, rowIndex, colIndex))
+	{
+		return false;
+	}
+
+	resultString = PQgetvalue(result, rowIndex, colIndex);
+
+	return pg_strtouint64(resultString, NULL, 10);
+}
+
+
+/*
+ * ParseTimestampTzField parses a timestamptz from a remote result or returns
+ * 0 if the result is NULL.
+ */
+static TimestampTz
+ParseTimestampTzField(PGresult *result, int rowIndex, int colIndex)
+{
+	char *resultString = NULL;
+	Datum resultStringDatum = 0;
+	Datum timestampDatum = 0;
+
+	if (PQgetisnull(result, rowIndex, colIndex))
+	{
+		return 0;
+	}
+
+	resultString = PQgetvalue(result, rowIndex, colIndex);
+	resultStringDatum = CStringGetDatum(resultString);
+	timestampDatum = DirectFunctionCall3(timestamptz_in, resultStringDatum, 0, -1);
+
+	return DatumGetTimestampTz(timestampDatum);
+}
+
+
+/*
+ * dump_local_wait_edges returns wait edges for distributed transactions
+ * running on the node on which it is called, which originate from the source node.
+ */
+Datum
+dump_local_wait_edges(PG_FUNCTION_ARGS)
+{
+	int32 sourceNodeId = PG_GETARG_INT32(0);
+
+	WaitGraph *waitGraph = BuildWaitGraphForSourceNode(sourceNodeId);
+	ReturnWaitGraph(waitGraph, fcinfo);
+
+	return (Datum) 0;
+}
+
+
+/*
+ * ReturnWaitGraph returns a wait graph for a set returning function.
+ */
+static void
+ReturnWaitGraph(WaitGraph *waitGraph, FunctionCallInfo fcinfo)
+{
+	ReturnSetInfo *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	TupleDesc tupdesc = NULL;
+	Tuplestorestate *tupstore = NULL;
+	MemoryContext per_query_ctx = NULL;
+	MemoryContext oldcontext = NULL;
+	size_t curEdgeNum = 0;
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg(
+					 "set-valued function called in context that cannot accept a set")));
+	}
+	if (!(rsinfo->allowedModes & SFRM_Materialize))
```

Comment:
Not doing this could cause crashes if the function is called from a context that doesn't support results returned using a tuple store, although I don't think such contexts exist right now. It's more a convention.